### PR TITLE
Loss functions and gradient fixes

### PR DESCRIPTION
**Pull request by AlexDBlack**
- Implements loss function methods (MSE, L1, L2, MCXENT so far)
- Adds gradient checks for transforms, more for reductions
- Gradient fixes for lots of transforms, implement some that were previously not implemented
- Gradient fixes for some transforms
- Adds OnesLike op
- A few extra utility methods, etc
  
Note that not all of these new tests are passing yet (I'll give them another pass later).

### CODE REVIEW

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/functions/DifferentialFunctionFactory.java [hunk 1]
```diff
@@ -173,37 +185,32 @@ public SDVariable normmax(SDVariable i_x, int... dimensions) {
...
-        validateDifferentialFunctionsameDiff(input);
-        SDVariable result;
-        if(Shape.isWholeArray(axes)) {
-            result = input;
-        }
-        else if(axes.length > 1) {
-            if(axes[0] > axes[1]) {
-                axes[0]--;
+    public SDVariable reductionBroadcastableWithOrigShape(int origRank, int[] reduceDims, SDVariable toExpand){
+        if(Shape.isWholeArray(origRank, reduceDims)){
```
Comment (agibsonccc) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/functions/DifferentialFunctionFactory.java]: May want to add a check to Shape.isWholeArray that covers zero length arrays as scalar results as well.

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/loss/LossFunctions.java [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+                                  Reduction reduction, int... dimensions){
+        LossInfo.Builder b = validate("mcxent", predictions, label, reduction);
+        SameDiff sd = predictions.getSameDiff();
+
+        if(weights == null){
+            weights = sd.one("mcxent_loss_weights", SCALAR);
+        }
+
+        String name = (reduction == Reduction.NONE ? outputName : null);
+        SDVariable weightedLogProd = sd.log(predictions).mul(label).mul(name, weights);
```
Comment (agibsonccc) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/loss/LossFunctions.java]: We have some in place optimizations that could be done here. Samediff supports in place ops as well.

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/accum/Max.java [hunk 3]
```diff
@@ -100,12 +100,17 @@ public float zeroFloat() {
...
         //TODO do we need to handle the "multiple equal maximums" case?
+        //TODO code duplication (min/max)
 
-        //Broadcast to input shape, work out IsMax (essentially)
-        SDVariable repeatOut = f().doRepeat(outputVariables()[0], arg());
-        SDVariable argMaxLoc = f().eq(repeatOut, arg());
-        SDVariable out = argMaxLoc.mul(i_v1.get(0));
-        return Collections.singletonList(out);
+        SDVariable out = outputVariables()[0];
+        int[] argShape = arg().getShape();
```
Comment (agibsonccc) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/accum/Max.java]: Arg may not always have a shape. We might have some place holder logic we want to put in here. At least checking for null shapes before you do this would be good.

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/accum/Min.java [hunk 4]
```diff
@@ -99,11 +99,18 @@ public String tensorflowName() {
...
         //TODO do we need to handle the "multiple equal minimums" case?
+        //TODO code duplication (min/max)
 
-        SDVariable repeatOut = f().doRepeat(outputVariables()[0], arg());
-        SDVariable argMaxLoc = f().eq(repeatOut, arg());
-        SDVariable out = argMaxLoc.mul(i_v1.get(0));
-        return Collections.singletonList(out);
+        SDVariable out = outputVariables()[0];
+
+        int[] argShape = arg().getShape();
```
Comment (agibsonccc) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/accum/Min.java]: Arg may not always have a shape. We might have some place holder logic we want to put in here. At least checking for null shapes before you do this would be good.

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/accum/NormMax.java [hunk 5]
```diff
@@ -79,7 +79,16 @@ public String opName() {
 
     @Override
     public List<SDVariable> doDiff(List<SDVariable> i_v1) {
-        SDVariable ret = f().doNormGrad(outputVariables()[0],i_v1.get(0),"normmax",dimensions);
+        //maxnorm(in) = max_i |x_i|
+        //d maxnorm(in)/dx = 0 if x_i is not the max, or d|x|/dx otherwise
+
+        SDVariable absIn = sameDiff.abs(arg());
+        SDVariable maxnorm = outputVariables()[0];
+        SDVariable maxnormBc = sameDiff.f().reductionBroadcastableWithOrigShape(arg().getShape().length, dimensions, maxnorm);    //TODO Is NPE possible on getShape()?
```
Comment (agibsonccc) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/accum/NormMax.java]: Arg may not always have a shape. We might have some place holder logic we want to put in here. At least checking for null shapes before you do this would be good.

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/accum/Prod.java [hunk 6]
```diff
@@ -100,7 +100,10 @@ public String tensorflowName() {
     @Override
     public List<SDVariable> doDiff(List<SDVariable> i_v1) {
         SDVariable prod = outputVariables()[0];
-        SDVariable ret = prod.div(arg()).mul(i_v1.get(0));
+        SDVariable broadcastableGrad = sameDiff.f().reductionBroadcastableWithOrigShape(arg().getShape().length, dimensions, i_v1.get(0));
```
Comment (agibsonccc) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/accum/Prod.java]: Arg may not always have a shape. We might have some place holder logic we want to put in here. At least checking for null shapes before you do this would be good.

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/ExpandDims.java [hunk 7]
```diff
@@ -48,6 +48,9 @@ public ExpandDims() {
 
     public ExpandDims(SameDiff sameDiff, SDVariable[] args, int axis) {
         super(null, sameDiff, args);
+        if(axis == Integer.MAX_VALUE){
```
Comment (agibsonccc) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/ExpandDims.java]: Maybe we want an Nd4jIllegalArgumentException?

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/Sqrt.java [hunk 8]
```diff
@@ -87,7 +88,9 @@ public String tensorflowName() {
 
     @Override
     public List<SDVariable> doDiff(List<SDVariable> i_v) {
-        throw new UnsupportedOperationException();
+        SDVariable out = arg();
+        SDVariable g = sameDiff.pow(out, -0.5).mul(0.5).mul(i_v.get(0));
```
Comment (agibsonccc) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/Sqrt.java]: In place optimizations here?

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/loss/LossFunctions.java (hunk 2 above)
Comment (AlexDBlack) [reply nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/loss/LossFunctions.java]: It's a reasonable point re: in-place ops, I'm wondering how we'll go with backprop with in-place though? Many ops require the original (unmodified) input.
I'm tempted to get things passing first (long way to go there yet), then do a separate optimization pass (rather than opening up that can of worms right now)

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/ExpandDims.java (hunk 7 above)
Comment (AlexDBlack) [reply nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/ExpandDims.java]: Done
